core/trino-parser/src/main/java/io/trino/sql/tree/Identifier.java

```diff
@@ -118,4 +128,17 @@ public boolean shallowEquals(Node other)
         Identifier that = (Identifier) other;
         return Objects.equals(value, that.value) && delimited == that.delimited;
     }
+
+    private static boolean isValidIdentifier(String value)
+    {
+        if (Strings.isNullOrEmpty(value)) {
+            return false;
+        }
+
+        if (!FIRST_CHAR_MATCHER.matches(value.charAt(0))) {
+            return false;
+        }
+
+        return TAIL_CHARS_MATCHER.matchesAllOf(value.substring(1));
```

Comment:
nit: `.substring` is copying operation. can we do better?

Reply:
Of course :)

Reply:
Better now?